File: sdks/python/glyph_protocol/verify.py
"""Signature verification — cards, receipts, manifests, key registries."""
from __future__ import annotations

from typing import Any, Optional

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PublicKey,
)

from .core import canonical_hash, fingerprint_key

# Fields that enter a card's canonical content (must match @glyphp/core).
_CARD_CANONICAL_FIELDS = [
    "version",
    "name",
    "intent",
    "tags",
    "cost",
    "idempotent",
    "input",
    "output",
    "examples",
    "failureModes",
    "provider",
    "attestation",
]


def _compute_glyph_id(card: dict) -> str:
    picked = {f: card.get(f) for f in _CARD_CANONICAL_FIELDS if f in card}
    return canonical_hash(picked)


def _verify_ed25519(public_key_hex: str, message: bytes, signature_hex: str) -> bool:
    try:
        pk = Ed25519PublicKey.from_public_bytes(bytes.fromhex(public_key_hex))
        pk.verify(bytes.fromhex(signature_hex), message)
        return True
    except (InvalidSignature, ValueError):
        return False


def verify_glyph(card: dict) -> bool:
    """Verifies the card's content hash and ed25519 signature."""
    pub = card.get("publicKey")
    sig = card.get("signature")
    if not pub or not sig:
        return False
    expected_id = _compute_glyph_id(card)
    if expected_id != card.get("id"):
        return False
    return _verify_ed25519(pub, expected_id.encode("ascii"), sig)


def verify_receipt(receipt: dict) -> bool:
    """Verifies a CallReceipt against its embedded serverPublicKey."""
    pub = receipt.get("serverPublicKey")
    sig = receipt.get("signature")
    if not pub or not sig:
        return False
    base = {k: v for k, v in receipt.items() if k != "signature"}
    return _verify_ed25519(pub, canonical_hash(base).encode("ascii"), sig)


def verify_manifest(manifest: dict) -> bool:
    """Verifies an UpdateManifest against its embedded serverPublicKey."""
    pub = manifest.get("serverPublicKey")
    sig = manifest.get("signature")
    if not pub or not sig:
        return False
    base = {k: v for k, v in manifest.items() if k != "signature"}
    return _verify_ed25519(pub, canonical_hash(base).encode("ascii"), sig)


# ---- key registry ---------------------------------------------------------

def _entry_payload(entry: dict) -> str:
    return canonical_hash(
        {
            "fingerprint": entry["fingerprint"],
            "publicKey": entry["publicKey"],
            "validFrom": entry["validFrom"],
            "signedBy": entry.get("signedBy"),
        }
    )


def _registry_payload(reg: dict) -> str:
    return canonical_hash(
        {
            "registryVersion": reg["registryVersion"],
            "serverId": reg["serverId"],
            "active": reg["active"],
            "keys": reg["keys"],
            "issuedAt": reg["issuedAt"],
            "ttlSeconds": reg["ttlSeconds"],
        }
    )


def verify_key_registry(registry: dict) -> bool:
    """Verifies a KeyRegistry's chain-of-trust and outer signature."""
    if registry.get("registryVersion") != "1.0":
        return False
    by_fp: dict[str, dict] = {}
    for entry in registry["keys"]:
        if fingerprint_key(entry["publicKey"]) != entry["fingerprint"]:
            return False
        by_fp[entry["fingerprint"]] = entry
    for entry in registry["keys"]:
        if not entry.get("signedBy"):
            continue
        sig = entry.get("signature")
        if not sig:
            return False
        parent = by_fp.get(entry["signedBy"])
        if not parent:
            return False
        if not _verify_ed25519(
            parent["publicKey"], _entry_payload(entry).encode("ascii"), sig
        ):
            return False
    active = by_fp.get(registry["active"])
    if not active or active.get("revokedAt"):
        return False
    return _verify_ed25519(
        active["publicKey"],
        _registry_payload(registry).encode("ascii"),
        registry["signature"],
    )


def resolve_key(registry: dict, public_key_hex: str) -> dict:
    """Returns `{status: 'active'|'retired'|'revoked'|'unknown', entry?, reason?}`."""
    target = fingerprint_key(public_key_hex)
    for entry in registry["keys"]:
        if entry["fingerprint"] != target:
            continue
        if entry.get("revokedAt"):
            return {
                "status": "revoked",
                "entry": entry,
                "reason": entry.get("revocationReason"),
            }
        if entry.get("validUntil"):
            return {"status": "retired", "entry": entry}
        return {"status": "active", "entry": entry}
    return {"status": "unknown"}

File: sdks/go/glyphprotocol/verify.go
package glyphprotocol

import (
	"crypto/ed25519"
	"encoding/hex"
	"errors"
)

// Verify a card's content hash and ed25519 signature. Returns (ok, error)
// — error is non-nil only for malformed inputs (hex decode failures);
// returning (false, nil) means the signature simply did not verify.
func VerifyGlyph(card map[string]any) (bool, error) {
	pub, _ := card["publicKey"].(string)
	sig, _ := card["signature"].(string)
	id, _ := card["id"].(string)
	if pub == "" || sig == "" || id == "" {
		return false, nil
	}
	expected, err := computeGlyphID(card)
	if err != nil {
		return false, err
	}
	if expected != id {
		return false, nil
	}
	return verifyEd25519(pub, []byte(expected), sig)
}

// VerifyReceipt verifies a CallReceipt against its embedded serverPublicKey.
func VerifyReceipt(receipt map[string]any) (bool, error) {
	pub, _ := receipt["serverPublicKey"].(string)
	sig, _ := receipt["signature"].(string)
	if pub == "" || sig == "" {
		return false, nil
	}
	base := copyWithout(receipt, "signature")
	hash, err := CanonicalHash(base)
	if err != nil {
		return false, err
	}
	return verifyEd25519(pub, []byte(hash), sig)
}

// VerifyManifest verifies an UpdateManifest against its embedded serverPublicKey.
func VerifyManifest(manifest map[string]any) (bool, error) {
	pub, _ := manifest["serverPublicKey"].(string)
	sig, _ := manifest["signature"].(string)
	if pub == "" || sig == "" {
		return false, nil
	}
	base := copyWithout(manifest, "signature")
	hash, err := CanonicalHash(base)
	if err != nil {
		return false, err
	}
	return verifyEd25519(pub, []byte(hash), sig)
}

// VerifyKeyRegistry verifies a registry's chain-of-trust and outer signature.
func VerifyKeyRegistry(registry map[string]any) (bool, error) {
	if v, _ := registry["registryVersion"].(string); v != "1.0" {
		return false, nil
	}
	rawKeys, _ := registry["keys"].([]any)
	if len(rawKeys) == 0 {
		return false, nil
	}
	byFp := map[string]map[string]any{}
	for _, raw := range rawKeys {
		entry, _ := raw.(map[string]any)
		pub, _ := entry["publicKey"].(string)
		fp, _ := entry["fingerprint"].(string)
		if FingerprintKey(pub) != fp {
			return false, nil
		}
		byFp[fp] = entry
	}
	for _, raw := range rawKeys {
		entry, _ := raw.(map[string]any)
		signedBy, _ := entry["signedBy"].(string)
		if signedBy == "" {
			continue
		}
		sig, _ := entry["signature"].(string)
		if sig == "" {
			return false, nil
		}
		parent, ok := byFp[signedBy]
		if !ok {
			return false, nil
		}
		payload, err := CanonicalHash(map[string]any{
			"fingerprint": entry["fingerprint"],
			"publicKey":   entry["publicKey"],
			"validFrom":   entry["validFrom"],
			"signedBy":    entry["signedBy"],
		})
		if err != nil {
			return false, err
		}
		ok2, err := verifyEd25519(parent["publicKey"].(string), []byte(payload), sig)
		if err != nil {
			return false, err
		}
		if !ok2 {
			return false, nil
		}
	}
	activeFp, _ := registry["active"].(string)
	active, ok := byFp[activeFp]
	if !ok {
		return false, nil
	}
	if _, revoked := active["revokedAt"]; revoked {
		return false, nil
	}
	payload, err := CanonicalHash(map[string]any{
		"registryVersion": registry["registryVersion"],
		"serverId":        registry["serverId"],
		"active":          registry["active"],
		"keys":            registry["keys"],
		"issuedAt":        registry["issuedAt"],
		"ttlSeconds":      registry["ttlSeconds"],
	})
	if err != nil {
		return false, err
	}
	outerSig, _ := registry["signature"].(string)
	return verifyEd25519(active["publicKey"].(string), []byte(payload), outerSig)
}

// ResolveResult mirrors @glyphp/core.ResolveResult.
type ResolveResult struct {
	Status string         // "active" | "retired" | "revoked" | "unknown"
	Entry  map[string]any // nil when Status == "unknown"
	Reason string         // populated when Status == "revoked"
}

// ResolveKey returns the registry's status for a given hex public key.
func ResolveKey(registry map[string]any, publicKeyHex string) ResolveResult {
	target := FingerprintKey(publicKeyHex)
	rawKeys, _ := registry["keys"].([]any)
	for _, raw := range rawKeys {
		entry, _ := raw.(map[string]any)
		fp, _ := entry["fingerprint"].(string)
		if fp != target {
			continue
		}
		if revoked, _ := entry["revokedAt"].(string); revoked != "" {
			reason, _ := entry["revocationReason"].(string)
			return ResolveResult{Status: "revoked", Entry: entry, Reason: reason}
		}
		if validUntil, _ := entry["validUntil"].(string); validUntil != "" {
			return ResolveResult{Status: "retired", Entry: entry}
		}
		return ResolveResult{Status: "active", Entry: entry}
	}
	return ResolveResult{Status: "unknown"}
}

// ---- internals -----------------------------------------------------------

var cardCanonicalFields = []string{
	"version", "name", "intent", "tags", "cost", "idempotent",
	"input", "output", "examples", "failureModes", "provider", "requiredScopes", "attestation",
}

func computeGlyphID(card map[string]any) (string, error) {
	picked := map[string]any{}
	for _, f := range cardCanonicalFields {
		if v, ok := card[f]; ok {
			picked[f] = v
		}
	}
	return CanonicalHash(picked)
}

func copyWithout(m map[string]any, key string) map[string]any {
	out := make(map[string]any, len(m))
	for k, v := range m {
		if k == key {
			continue
		}
		out[k] = v
	}
	return out
}

func verifyEd25519(publicKeyHex string, message []byte, signatureHex string) (bool, error) {
	pubBytes, err := hex.DecodeString(publicKeyHex)
	if err != nil {
		return false, err
	}
	if len(pubBytes) != ed25519.PublicKeySize {
		return false, errors.New("ed25519: invalid public key length")
	}
	sigBytes, err := hex.DecodeString(signatureHex)
	if err != nil {
		return false, err
	}
	if len(sigBytes) != ed25519.SignatureSize {
		return false, nil
	}
	return ed25519.Verify(ed25519.PublicKey(pubBytes), message, sigBytes), nil
}
